### DISCUSSION TITLE
I18n in Rails

### DISCUSSION THREAD

**Discussion by dhnaranjo** [yippee-fun/phlex#577] [Ideas]
Whattup y'all. So, phlex-translate exists, but in a Rails app I'd like to take advantage of the built in translation inference and lazy lookup biz as much as possible. Rails translate is accessible through `helpers`, but if you hit `helpers.translate(".some_scoped_key")` you get `Cannot use t(".some_scoped_key") shortcut because path is not available`. The unavailable path comes from the `@virtual_path` instance variable on the view context. Anyways, I ended up here:

```
class ApplicationComponent < Phlex::HTML
  ...
  def t(...)
    helpers.instance_variable_set(:@virtual_path, translation_path)
    helpers.translate(...)
  end

  def translation_path
    @_parent&.translation_path || helpers.controller_path.tr("/", ".")
  end
  ...
end

class ApplicationView < ApplicationComponent
  def translation_path
    path_name = self.class.name.demodulize.underscore.delete_suffix("_view")
    "#{super()}.#{path_name}"
  end
end
```

This allows me to make components like:
```
class ThingComponent < ApplicationComponent
  def template
    t(".thing", default: t("components.thing"))
  end
end

Businesses::ShowView < ApplicationView
  def template
    render(ThingComponent)
  end
end

BusinessController < ApplicationController
  def show
    render ShowView.new
  end
end
```

Now when Thing goes to render it'll first look at "businesses.show.thing" before falling back to "components.thing", allowing me to have an overrideable default where-ever I use the component.

And lemme tell you it works out real nice. Now, other than that setting an instance variable on another object directly is gross, is anything wrong with this approach?

**Comment by davidalejandroaguilar** [comment]
Isn't this solved by using [phlex-translation](https://github.com/phlex-ruby/phlex-translation) as [the docs mention](https://www.phlex.fun/translations/)?

The only caveat is that if you have `Admin::Users::ShowView` your translation key must be `en.admin.users.showview` (instead of `show_view`), which is not ideal, but it works.

**Reply by stex** [reply]
Which can be solved by overriding the translation path:

```ruby
def self.translation_path
  @translation_path ||= name&.split("::")&.join(".")&.underscore.to_s
end
```

In the end, `phlex-translation` is only 2 methods
